src/Interpreters/executeQuery.cpp

```diff
@@ -1085,7 +1086,8 @@ void executeQuery(
                     out->onProgress(progress);
                 });
 
-                out->setBeforeFinalizeCallback(before_finalize_callback);
+                if (flush_buffer_func)
```

Comment:
Seems that this callback is needed only to update the progress?
Maybe this callback can be replaced by simply returning number of processed/result rows?
What do you think?

Reply:
> Seems that this callback is needed only to update the progress?

Not only. As important is also to clear previous progress output to avoid mess between query output and progress output. (also see mentioned it here https://github.com/ClickHouse/ClickHouse/pull/26231#discussion_r680553682)